src/index.js

```diff
@@ -1,37 +1,79 @@
+// we use requestAnimationFrame to be called by the browser before every repaint
+let requestAnimationFrame;
+
 // Get the top position of an element in the document
 const getTop = function(element, start) {
   // return value of html.getBoundingClientRect().top ... IE : 0, other browsers : -pageYOffset
   if (element.nodeName === 'HTML') return -start;
   return element.getBoundingClientRect().top + start;
 };
 
+const smoothScrollCtx = Symbol('smoothScrollCtx')
+
+export const smoothScroll = function ({ scrollTo, offset = 0, duration = 500, container = window, updateHistory = true, hash }) {
+  if (!requestAnimationFrame) {
+    requestAnimationFrame = window.requestAnimationFrame ||
+    function(fn) {
+      window.setTimeout(fn, 16);
+    };
+  }
+  
+  // Using the history api to solve issue: back doesn't work
+  // most browser don't update :target when the history api is used:
+  // THIS IS A BUG FROM THE BROWSERS.
+  if (updateHistory && window.history.pushState && location.hash !== hash) window.history.pushState('', '', hash);
+
+
+  const startPoint = container.scrollTop || window.pageYOffset;
+  // Get the top position of an element in the document
+  // return value of html.getBoundingClientRect().top ... IE : 0, other browsers : -pageYOffset
+  let end = getTop(scrollTo, startPoint);
+
+  // Ajusts offset from the end
+  end += offset;
+
+  const clock = Date.now();
+  const step = function() {
+    // the time elapsed from the beginning of the scroll
+    const elapsed = Date.now() - clock;
+    // calculate the scroll position we should be in
+    let position = end;
+    if (elapsed < duration) {
+      position = startPoint + (end - startPoint) * easeInOutCubic(elapsed / duration);
+
+      requestAnimationFrame(step);
+    } else if (updateHistory) {
+      location.replace('#' + scrollTo.id);
+      // this will cause the :target to be activated.
+    }
+
+    container === window ? container.scrollTo(0, position) : (container.scrollTop = position);
+  };
+  step();
+}
+
 const VueSmoothScroll = {
   install(Vue, config) {
+    const defaultValue = {
+      duration: 500,
+      offset: 0,
+      container: window,
+      updateHistory: true,
+    };
+
     Vue.directive('smooth-scroll', {
-      inserted(el, binding) {
+      inserted(el, binding, vnode) {
         // Do not initialize smoothScroll when running server side, handle it in client
         // We do not want this script to be applied in browsers that do not support those
         // That means no smoothscroll on IE9 and below.
         if (typeof window !== 'object' || window.pageYOffset === undefined) return;
 
-        const defaultValue = {
-          duration: 500,
-          offset: 0,
-          container: window,
-          updateHistory: true,
-        };
+        const hash = vnode.data.attrs.href;
```

Comment:
I just realized that this is not your change, but were made in 01ed33483b9d9403b165baf027bcbc68169fd639.

There's something up with the rebase, or these changes would not show up. Can you please try squashing your commits, and force pushing on top of this PR branch (which should be on top of latest master)?

Reply:
For the issue with dynamic href, it may be better to move this back in the click handler. @Yuliang-Lee?

Reply:
Sorry about the messy rebase - i've done the squash like you suggested so hopefully that makes things clearer.

Reply:
> For the issue with dynamic href, it may be better to move this back in the click handler. @Yuliang-Lee?

It looks you are right, I will move this to click handler